### ISSUE TITLE
Improve support for ARIA focus

### ISSUE THREAD

**Issue by tjfred35** [gpbl/react-day-picker#1278]
The example in https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/datepicker-dialog.html for keyboard navigation suggests that focus should move to the current date (today) or the selected date when opening a date picker dialog

What’s more, the date that was previously selected with the keyboard is preserved when tabbing backwards and forwards to other focusable elements.

The logic could be:
- track the users interaction. if the user’s previously focused day is not disabled and in the current month, focus on that day
- if a non-disabled selected day is in the current month, focus on the first non-disabled selected day 
- If today is in the current month (and not disabled) focus on today
- otherwise focus on the first non-disabled day of the month

Additionally, there needs to be some way to focus on the day when the DayPicker first renders, or re-opens in the modal Dialog.

**Comment by tjfred35** [comment]
I have a working prototype for this, FWIW

**Comment by gpbl** [comment]
@tjfred35 prototypes welcome, but we must add the tests first to the current code.

**Comment by gpbl** [comment]
@tjfred35 

> The example in https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/datepicker-dialog.html for keyboard navigation suggests that focus should move to the current date (today) or the selected date when opening a date picker dialog

I'm missing something here: should the keyboard navigation change when DayPicker is in a dialog?

> What’s more, the date that was previously selected with the keyboard is preserved when tabbing backwards and forwards to other focusable elements.

What do you mean that is preserved?

**Comment by jpveooys** [comment]
We are also building a date picker using react-day-picker that functions similarly to https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/datepicker-dialog.html.

> I'm missing something here: should the keyboard navigation change when DayPicker is in a dialog?

Yes – the calendar button is opening a modeless dialog containing a date picker in that example. Arguably react-day-picker doesn't know when it's being used in this fashion, but I can't see a clean way to achieve the desired focusing behaviour either.

> What do you mean that is preserved?

@tjfred35 may to referring to the focus currently going back to the first day in the month when cycling through tabbable elements using tab, rather than the last day that was focused. (Note also that the linked example uses a focus trap when the date picker is open, so the tabbing behaviour is more obvious there.)

**Comment by gpbl** [comment]
Gotcha thanks! I guess these are more than one a11y features:

- [ ] When the dialog is opened, if the input field is empty, or does not contain a valid date, then the current date is focused in the calendar. Otherwise, the focus is placed on the day in the calendar that matches the value of the date input field.
- [ ] When a date is chosen, the accessible name of the Choose Date button is updated to include the selected date. So, when the dialog closes and focus returns to the Choose Date button, screen reader users hear confirmation of the selected date in the button name.
- [ ] When the month or year of the calendar grid changes as users navigate the calendar or activate the buttons for next or previous month or year, a live region enables screen readers to announce the new month and year.
- [ ] The calendar grid provides hotkeys for changing the year and month as well as support for normal grid navigation keys.
When the dialog opens and a date in the calendar grid receives focus, a live region enables screen readers to announce keyboard instructions for navigating the calendar grid. The instructions are also visible at the bottom of the dialog box.

I think we should actually implement DayPicker in a dialog like in the example above, and from there see what should be supported by DayPicker and what instead should go to the dialog implementation. What do you all think?

**Comment by gpbl** [comment]
> When the dialog is opened, if the input field is empty, or does not contain a valid date, then the current date is focused in the calendar. Otherwise, the focus is placed on the day in the calendar that matches the value of the date input field.

Maybe this could be easily done with the FocusContext and a new `focusedDay` prop.

**Comment by jpveooys** [comment]
> I think we should actually implement DayPicker in a dialog like in the example above, and from there see what should be supported by DayPicker and what instead should go to the dialog implementation. What do you all think?

Yes, that seems like a sensible way forward to me.

**Comment by gpbl** [comment]
Closing as #1343 has been merged.
